### DISCUSSION TITLE
Cache based on cell content

### DISCUSSION THREAD

**Discussion by akrogp** [quarto-dev/quarto-cli#7491] [Q&A] [answered]
### Description

I usually enable caching of long running code chunks but, If I insert another code chunk before, the chunk (and the following ones) are evaluated again despite their contents have not changed.

I guess that code chunks are identified by their position and thus content change is checked in a position-specific basis. Would it be possible to identify code chunks using a content hash or similar so that changing the order of chunks does not trigger their re-evaluation?

Thanks

**Comment by mcanouil** [comment]
Hi, caching occurs differently between Jupyter and knitr, so as-is there is no concrete answer to your question.
Could you share a small self-contained "working" (reproducible) example to work with, _i.e._, a complete Quarto document or a Git repository? Thanks.
Please note that caching is not handled by Quarto, so there is little that can be done by Quarto team in this matter.

You can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Reply by akrogp** [reply]
Sure, sorry for not including more details. I am using RStudio with R code chunks. Here there is an example:

````qmd
---
title: "Testing code chunk caching from RStudio"
format: html
---

If I insert a new code chunk here, "Chunk 2" is re-evaluated although it has not been changed.

```{r}
# Chunk 1
a <- 2
```

```{r}
#| cache: true

# Chunk 2
a + 1
```
````

**Reply by mcanouil** [reply]
So indeed, this is a knitr behaviour that cannot be changed on Quarto side.
If you want more granular control, I suggest to look up `targets` for handling your workflow, caching, code cells dependencies.

**Reply by akrogp** [reply, marked as answer]
Thanks for the suggestion. I have just tried assigning a label to the code block and now caching works despite of chunks reordering, so it is also a good workaround.

**Reply by mcanouil** [reply]
Oh yes, knitr auto labels code cells if none are provided.
So adding code celles changes the number used on the label thus breaking link with the cache which is stored based on the labels.

**Comment by akrogp** [comment]
I close the discussion since it is a knitr behaviour and using labels serves as a workaround.
